File: sdks/js/packages/spark-sdk/src/utils/transfer_package.ts
import { hexToBytes } from "@noble/curves/abstract/utils";
import { sha256 } from "@noble/hashes/sha2";
import { TransferPackage } from "../proto/spark.js";

// GetTransferPackageSigningPayload returns the signing payload for a transfer package.
// The payload is a hash of the transfer ID and the encrypted payload sorted by key.
export function getTransferPackageSigningPayload(
  transferID: string,
  transferPackage: TransferPackage,
): Uint8Array {
  const encryptedPayload = transferPackage.keyTweakPackage;

  // convert map to array of key-value pairs
  const pairs: { key: string; value: Uint8Array }[] = Object.entries(
    encryptedPayload,
  ).map(([key, value]) => ({ key, value }));

  // Sort the slice by key to ensure deterministic ordering
  // This is important for consistent signing payloads
  pairs.sort((a, b) => a.key.localeCompare(b.key));

  const encoder = new TextEncoder();
  let message = hexToBytes(transferID.replaceAll("-", ""));

  for (const pair of pairs) {
    const keyPart = encoder.encode(pair.key + ":");
    const separator = encoder.encode(";");

    message = new Uint8Array([
      ...message,
      ...keyPart,
      ...pair.value,
      ...separator,
    ]);
  }

  return sha256(message);
}

File: spark/common/transfer_package.go
package common

import (
	"cmp"
	"crypto/sha256"
	"slices"

	"github.com/google/uuid"
	pb "github.com/lightsparkdev/spark/proto/spark"
)

// GetTransferPackageSigningPayload returns the signing payload for a transfer package.
// The payload is a hash of the transfer ID and the encrypted payload sorted by key.
func GetTransferPackageSigningPayload(transferID uuid.UUID, transferPackage *pb.TransferPackage) []byte {
	encryptedPayload := transferPackage.KeyTweakPackage
	// Create a slice to hold the sorted key-value pairs
	type keyValuePair struct {
		key   string
		value []byte
	}

	// Convert map to slice of key-value pairs
	pairs := make([]keyValuePair, 0, len(encryptedPayload))
	for k, v := range encryptedPayload {
		pairs = append(pairs, keyValuePair{key: k, value: v})
	}

	// Sort the slice by key to ensure deterministic ordering
	// This is important for consistent signing payloads
	slices.SortFunc(pairs, func(a, b keyValuePair) int { return cmp.Compare(a.key, b.key) })

	hasher := sha256.New()
	hasher.Write(transferID[:])
	for _, pair := range pairs {
		hasher.Write([]byte(pair.key + ":"))
		hasher.Write(pair.value)
		hasher.Write([]byte(";"))
	}

	return hasher.Sum(nil)
}
